### DISCUSSION TITLE
Overriding DynamicResource Brush

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#13517] [Q&A] [answered]
I'm currently trying to override some default brushes used in an external dependency, and having some issues with the process.
Apparently, the only thing i should have to do is declare another brush with the same key, but that does not seem to b the case?

Probably is due to my case, being the following : 

The brushes i want to override are placed in a dialog's xaml file as follow (remember that this dialog is contained in an external dependency)
```axaml
<Window.Resources>
  <SolidColorBrush x:Key="BindingEditorDialogBorderColorBrush" Color="#FFFFFF" />
  <SolidColorBrush x:Key="BindingEditorDialogBorderBackgroundBrush" Color="#1E2738" />
  <SolidColorBrush x:Key="BindingEditorDialogBodyColorBrush" Color="#171C2C" />
  <SolidColorBrush x:Key="BindingEditorDialogForegroundBrush" Color="#FFFFFF" />
</Window.Resources>
```

In my app's xaml, i'm including a resource dictionary this way 
```xaml
<Application.Resources>
  <ResourceDictionary>
    <ResourceDictionary.MergedDictionaries>
      <ResourceInclude Source="/Resources/BindingEditorDialogResources.axaml" />
    </ResourceDictionary.MergedDictionaries>

    ...

  </ResourceDictionary>
</Application.Resources>
```
`BindingEditorDialogResources.axaml` contains the following
```axaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">

  <SolidColorBrush x:Key="BindingEditorDialogBorderColorBrush" Color="#FFFFFF" />
  <SolidColorBrush x:Key="BindingEditorDialogBorderBackgroundBrush" Color="{StaticResource AccentColorBrush}" />
  <SolidColorBrush x:Key="BindingEditorDialogBodyColorBrush" Color="{StaticResource BodyColorBrush}" />
  <SolidColorBrush x:Key="BindingEditorDialogForegroundBrush" Color="#FFFFFF" />

</ResourceDictionary>
```
            
The only thing i can think of is that i would have to create a whole theme file just for those dialogs, and include it or not in the same way as fluent or simple themes.

**Comment by maxkatz6** [comment, marked as answer]
Two possible problems:
1. Your BindingEditorDialogResources.axaml is defined BEFORE external dependency resources. I.e. external resources simply override yours. This is most likely.
2. External dependency uses these resources with StaticResource and reads resources BEFORE you set your BindingEditorDialogResources.axaml. This is unlikely.

**Reply by Mrcubix** [reply]
> Your BindingEditorDialogResources.axaml is defined BEFORE external dependency resources. I.e. external resources simply override yours. This is most likely.

So the order would be App.axaml -> anything else, since the default resources are currently contained within the dialog file itself, this would make sense.

> External dependency uses these resources with StaticResource and reads resources BEFORE you set your BindingEditorDialogResources.axaml. This is unlikely.

The resources are used with DynamicResources

Would there be an easy way to achieve what i want? Or i have to go through the potential suffering of making a theme just for those elements and proceed like or similar to what i previously specified?

**Reply by maxkatz6** [reply]
> So the order would be App.axaml -> anything else, since the default resources are currently contained within the dialog file itself, this would make sense.

See resources resolution order https://docs.avaloniaui.net/docs/styling/resources#resource-resolution

If dialog control has defined resources directly on that control, it would be pretty hard to redefine them. 
That's why normally any control resources/styles are included in the Application.xaml instead of merging them into Control.Resources/Styles. So it can be ordered and redefined.

> Would there be an easy way to achieve what i want?

Is this dialog a UserControl or TemplatedControl? Either way, if I understood situation correctly, you can do something like this:
```
var control = new DialogControl();
control.Resources.MergedResources(new StyleInclude { Source = new Uri("avares://PathTo/BindingEditorDialogResources.xaml") };
```

Still, it depends on how DialogControl was written. 
The "correct" way would be to change that control instead.

**Reply by Mrcubix** [reply]
- The dialog is a `Window`,
- But i guess i'll try to work out something around your suggestion

**Reply by Mrcubix** [reply]
Figured that using `Color="{StaticResource AccentColorBrush}"` won't work, since, AccentColorBrush, is, a `SolidColorBrush`.
Will need to define static `Color` in my style files and use them instead.

**Reply by maxkatz6** [reply]
```
<StaticResource x:Key="BindingEditorDialogBorderBackgroundBrush" ResourceKey="AccentColorBrush" />
```
might work, assuming AccentColorBrush is a single instance for any theme, and it's Color value is dynamically binded. 
But yes, having AccentColor instead might be a cleaner solution.
